### PR TITLE
Pipeline Docs

### PR DESCRIPTION
**Pull request by hadarhg**
- [x] Design Your Pipeline
- [x] Create Your Pipeline
- [x] Test Your Pipeline

### CODE REVIEW

**Comment by hadarhg** [comment]
R: @francesperry

File: src/documentation/pipelines/create-your-pipeline.md [hunk 1]
```diff
@@ -5,4 +5,153 @@ permalink: /documentation/pipelines/create-your-pipeline/
...
+
+// Then create the pipeline.
+Pipeline p = Pipeline.create(options);
+```
+
+### Configuring Pipeline Options
+
+Use the pipeline options to configure different aspects of your pipeline, such as the pipeline runner that will execute your pipeline and any runner-specific configuration required by the chosen runner.
+
+For example, if you're using the Google Cloud Dataflow service to run your pipeline, your pipeline options will include information about your Cloud Platform project required by the Cloud Dataflow service, such as your project ID and Cloud Storage staging locations. The pipeline options also let you control how many workers the Dataflow service should assign to your pipeline job, and where to direct your pipeline job's status messages.
```
Comment (francesperry) [inline src/documentation/pipelines/create-your-pipeline.md]: Rephrase to use a non-Dataflow example.

File: src/documentation/pipelines/create-your-pipeline.md [hunk 2]
```diff
@@ -5,4 +5,153 @@ permalink: /documentation/pipelines/create-your-pipeline/
...
+
+A key property in the pipeline options that determines where your pipeline executes is the pipeline runner. The pipeline runner property also specifies whether your pipeline's execution is to be asynchronous or blocking.
```
Comment (francesperry) [inline src/documentation/pipelines/create-your-pipeline.md]: Don't think this is true any more given waitToFinish(). Cut paragraph?

File: src/documentation/pipelines/create-your-pipeline.md [hunk 3]
```diff
@@ -5,4 +5,153 @@ permalink: /documentation/pipelines/create-your-pipeline/
...
+
+When you run the pipeline on a runner of your choice, a copy of the PipelineOptions will be available to your code. For example, you can read PipelineOptions from a DoFn's Context.
+
+> **Note:** Access `PipelineOptions` inside any [ParDo](/documentation/programming-guide/#transforms-pardo)'s `DoFn` instance by using the method `ProcessContext.getPipelineOptions`.
+
+#### Setting PipelineOptions from Command-Line Arguments
+
+While you can configure your pipeline by creating a `PipelineOptions` object and setting the fields directly, the Beam SDKs include a command-line parser that you can use to set fields in `PipelineOptions` using command-line arguments.
+
+To read options from the command-line, construct your `PipelineOptions` object using the method `PipelineOptionsFactory.fromArgs`, as in the following example code:
```
Comment (francesperry) [inline src/documentation/pipelines/create-your-pipeline.md]: the formatting of fromArgs is going to be language-dependent. remove from prose?

File: src/documentation/pipelines/create-your-pipeline.md [hunk 4]
```diff
@@ -5,4 +5,153 @@ permalink: /documentation/pipelines/create-your-pipeline/
...
+
+For example, if you're using the Google Cloud Dataflow service to run your pipeline, your pipeline options will include information about your Cloud Platform project required by the Cloud Dataflow service, such as your project ID and Cloud Storage staging locations. The pipeline options also let you control how many workers the Dataflow service should assign to your pipeline job, and where to direct your pipeline job's status messages.
+
+A key property in the pipeline options that determines where your pipeline executes is the pipeline runner. The pipeline runner property also specifies whether your pipeline's execution is to be asynchronous or blocking.
+
+When you run the pipeline on a runner of your choice, a copy of the PipelineOptions will be available to your code. For example, you can read PipelineOptions from a DoFn's Context.
+
+> **Note:** Access `PipelineOptions` inside any [ParDo](/documentation/programming-guide/#transforms-pardo)'s `DoFn` instance by using the method `ProcessContext.getPipelineOptions`.
```
Comment (francesperry) [inline src/documentation/pipelines/create-your-pipeline.md]: formatting of getPipelineOptions will be language dependent.

File: src/documentation/pipelines/create-your-pipeline.md [hunk 5]
```diff
@@ -5,4 +5,153 @@ permalink: /documentation/pipelines/create-your-pipeline/
...
+
+To create your pipeline's initial `PCollection`, you apply a root transform to your pipeline object. A root transform creates a `PCollection` from either an external data source or some local data you specify.
+
+There are two kinds of root transforms in the Beam SDKs: `Read` and `Create`. `Read` transforms read data from an external source, such as a text file or a database table. `Create` transforms create a `PCollection` from an in-memory `java.util.Collection`.
+
+The following example code shows how to `apply` a `TextIO.Read` root transform to read data from a text file. The transform is applied to a `Pipeline` object `p`, and returns a pipeline data set in the form of a `PCollection<String>`:
+
+```java
+PCollection<String> lines = p.apply(
+  TextIO.Read.named("ReadMyFile").from("gs://some/inputData.txt"));
```
Comment (francesperry) [inline src/documentation/pipelines/create-your-pipeline.md]: .named is gone in Beam entirely.

.apply("ReadLines", TextIO.Read.from("gs://some/inputData.txt"))

File: src/documentation/pipelines/create-your-pipeline.md [hunk 6]
```diff
@@ -5,4 +5,153 @@ permalink: /documentation/pipelines/create-your-pipeline/
...
+## Writing or Outputting Your Final Pipeline Data
+
+Once your pipeline has applied all of its transforms, you'll usually need to output the results. To output your pipeline's final `PCollection`s, you apply a `Write` transform to that `PCollection`. `Write` transforms can output the elements of a `PCollection` to an external data sink, such as a database table. You can use `Write` to output a `PCollection` at any time in your pipeline, although you'll typically write out data at the end of your pipeline.
+
+The following example code shows how to `apply` a `TextIO.Write` transform to write a `PCollection` of `String` to a text file:
+
+```java
+PCollection<String> filteredWords = ...;
+
+filteredWords.apply(TextIO.Write.named("WriteMyFile").to("gs://some/outputData.txt"));
```
Comment (francesperry) [inline src/documentation/pipelines/create-your-pipeline.md]: .named is gone. use a first arguement to apply

File: src/documentation/pipelines/create-your-pipeline.md [hunk 7]
```diff
@@ -5,4 +5,153 @@ permalink: /documentation/pipelines/create-your-pipeline/
...
+```
+
+## Running Your Pipeline
+
+Once you have constructed your pipeline, you use the `run` method to execute the pipeline. Pipelines are executed asynchronously: the program you create sends a specification for your pipeline to a **pipeline runner**, which then constructs and runs the actual series of pipeline operations. 
+
+```java
+p.run();
+```
```
Comment (francesperry) [inline src/documentation/pipelines/create-your-pipeline.md]: Introduce .waitUntilFinish();

Comment (francesperry) [inline src/documentation/pipelines/create-your-pipeline.md]: Link to "Design" as the next thing to read, like you did previously?

File: src/documentation/pipelines/test-your-pipeline.md [hunk 8]
```diff
@@ -5,4 +5,263 @@ permalink: /documentation/pipelines/test-your-pipeline/
 ---
 # Test Your Pipeline
 
-> **Note:** There is an open JIRA issue to create this guide ([BEAM-901](https://issues.apache.org/jira/browse/BEAM-901)).
+* TOC
+{:toc}
+
+Testing your pipeline is a particularly important step in developing an effective data processing solution. The indirect nature of the Beam model, in which your user code constructs a pipeline graph to be executed remotely, can make debugging-failed runs a non-trivial task. Often it is faster and simpler to perform local unit testing on your pipeline code than to debug a pipeline's remote execution.
+
+Before running your pipeline on the service of your choice, unit testing your pipeline code locally is often the best way to identify and fix bugs in your pipeline code. Unit testing your pipeline locally also allows you to use your familiar/favorite local debugging tools.
```
Comment (francesperry) [inline src/documentation/pipelines/test-your-pipeline.md]: service -> runner

File: src/documentation/pipelines/test-your-pipeline.md [hunk 9]
```diff
@@ -5,4 +5,263 @@ permalink: /documentation/pipelines/test-your-pipeline/
 ---
 # Test Your Pipeline
 
-> **Note:** There is an open JIRA issue to create this guide ([BEAM-901](https://issues.apache.org/jira/browse/BEAM-901)).
+* TOC
+{:toc}
+
+Testing your pipeline is a particularly important step in developing an effective data processing solution. The indirect nature of the Beam model, in which your user code constructs a pipeline graph to be executed remotely, can make debugging-failed runs a non-trivial task. Often it is faster and simpler to perform local unit testing on your pipeline code than to debug a pipeline's remote execution.
+
```
Comment (francesperry) [inline src/documentation/pipelines/test-your-pipeline.md]: Introduce the DirectRunner (with link to the page Melissa is adding) by name as something great for testing and local development?

File: src/documentation/pipelines/test-your-pipeline.md [hunk 10]
```diff
@@ -5,4 +5,263 @@ permalink: /documentation/pipelines/test-your-pipeline/
...
+Before running your pipeline on the service of your choice, unit testing your pipeline code locally is often the best way to identify and fix bugs in your pipeline code. Unit testing your pipeline locally also allows you to use your familiar/favorite local debugging tools.
+
+**Note:** When testing your code using the Google Cloud Dataflow service, consider limiting the number of worker instances for your pipeline to 1, or the minimum number appropriate for your test. Limiting the number of worker instances used during repeated test runs can provide significant time and cost savings. You can limit the number of workers your pipeline uses during test runs by setting the `--maxNumWorkers` execution option when you run your test pipeline.
```
Comment (francesperry) [inline src/documentation/pipelines/test-your-pipeline.md]: Replace with a comment about how after you develop against the DirectRunner, you'll like want to use your runner of choice to test on a small scale. For example the Flink runner can be used with a local Flink cluster or a remote Flink cluster.

File: src/documentation/pipelines/test-your-pipeline.md [hunk 11]
```diff
@@ -5,4 +5,263 @@ permalink: /documentation/pipelines/test-your-pipeline/
...
+
+The code in your pipeline's `DoFn` functions runs often, and often across multiple Compute Engine instances. Unit-testing your `DoFn` objects before running them using a runner service can save a great deal of debugging time and energy.
+
+The Beam SDK for Java provides a convenient way to test an individual `DoFn` called [DoFnTester](https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/DoFnTesterTest.java), which is included in the SDK `Transforms` package.
+
+`DoFnTester`uses the [JUnit](http://junit.org) framework. To use `DoFnTester`, you'll need to do the following:
+
+1.  Create a `DoFnTester`. You'll need to pass an instance of the `DoFn` you want to test to the static factory method for `DoFnTester`.
+2.  Create one or more main test inputs of the appropriate type for your `DoFn`. If your `DoFn` takes side inputs and/or produces side outputs, you should also create the side inputs and the side output tags.
+3.  Call `DoFnTester.processBatch` to process the main inputs.
```
Comment (francesperry) [inline src/documentation/pipelines/test-your-pipeline.md]: processBundle?

(Would be good to confirm these code snippets.)

File: src/documentation/pipelines/test-your-pipeline.md [hunk 12]
```diff
@@ -5,4 +5,263 @@ permalink: /documentation/pipelines/test-your-pipeline/
...
+
+For tests, use `TestPipeline` in place of `Pipeline` when you create the pipeline object. Unlike `Pipeline.create`, `TestPipeline.create` handles setting `PipelineOptions` interally.
+
+You create a `TestPipeline` as follows:
+
+```java
+Pipeline p = TestPipeline.create();
+```
+
+#### PAssert
```
Comment (francesperry) [inline src/documentation/pipelines/test-your-pipeline.md]: Let's merge in a lot of @tgroh's blog contents?

Ok to leave a link to http://beam.incubator.apache.org/blog/2016/10/20/test-stream.html for now and file an issue to add it later.

File: src/documentation/pipelines/test-your-pipeline.md [hunk 13]
```diff
@@ -5,4 +5,263 @@ permalink: /documentation/pipelines/test-your-pipeline/
...
+    <groupId>org.hamcrest</groupId>
+    <artifactId>hamcrest-all</artifactId>
+    <version>1.3</version>
+    <scope>test</scope>
+</dependency>
+```
+
+For more information on how these classes work, see the [org.apache.beam.sdk.testing](http://beam.incubator.apache.org/documentation/sdks/javadoc/0.3.0-incubating/org/apache/beam/sdk/testing/package-summary.html) package documentation.
+
+### Using the Create Transform
```
Comment (francesperry) [inline src/documentation/pipelines/test-your-pipeline.md]: Feels like this should be before PAssert? So 
(1) create a test pipeline
(2) give it data
(3) hook up your composite
(4) passert the result

**Comment by davorbonaci** [comment]
Seems like there are two copies of the same file: one in `src`, one outside. I think we need to merge them.

File: src/documentation/pipelines/create-your-pipeline.md [hunk 14]
```diff
@@ -5,4 +5,157 @@ permalink: /documentation/pipelines/create-your-pipeline/
...
+
+You can also specify a description, which appears when a user passes `--help` as a command-line argument, and a default value.
+
+You set the description and default value using annotations, as follows:
+
+```java
+public interface MyOptions extends PipelineOptions {
+    @Description("My custom command line argument.")
+    @Default.String("D-FAULT")
```
Comment (davorbonaci) [inline src/documentation/pipelines/create-your-pipeline.md]: DEFAULT?

File: src/documentation/pipelines/design-your-pipeline.md [hunk 15]
```diff
@@ -5,4 +5,110 @@ permalink: /documentation/pipelines/design-your-pipeline/
...
+         alt="A pipeline with multiple input sources.">
+</figure>
+Figure 5: A pipeline with multiple input sources.
+
+## What's next
+
+*   [Create your own pipeline]({{ site.baseurl }}/documentation/pipelines/create-your-pipeline).
+*   [Test your pipeline]({{ site.baseurl }}/documentation/pipelines/test-your-pipeline).
+
+
```
Comment (davorbonaci) [inline src/documentation/pipelines/design-your-pipeline.md]: remove extra newlines here

**Comment by davorbonaci** [comment]
LGTM. Merging.

I'm sure we could iterate on this more, and there are likely outstanding issues. However, this is so much better than what we currently have that we should merge right away.

**Pull request closed without merging**